_posts/2021/2021-09-28-cpp-lessons-learned.md

```diff
@@ -0,0 +1,143 @@
+We wanted to do the hard work once, test it as much as possible, and give people reliable,
+efficient, ready-to-use solutions. We describe below only a few of our problems, which
+we believe might benefit for a broader audience.
+
+* **Optimize number of allocations!**
+
+Lots of allocations on pmem may ruin performance. It seems like an obvious statement, but
+it's easy to overlook this issue. One badly placed allocation (e.g., in critical
+code path), and we end up with low performance. Unfortunately, inserting a new node in
+`pmem::obj::concurrent_hash_map` is associated with multiple allocations.
+When we were thinking about the next container to implement, we chose radix tree structure.
+It is a great fit for pmem, because it requires a minimal number of allocations, no tree
+re-balancing, and it is sorted but requires no key comparisons.
+We realized `pmem::obj::experimental::radix_tree` can be optimized even further.
+In (probably) the most common scenario - key, value, or both are of type `string`
+(`pmem::obj::string` to be precise). Each such string actually equals two allocations.
+
+<figure class="image">
+  <img src="/assets/cpp_lessons_string.png" alt="Radix tree with string">
+  <figcaption>Figure 1. Radix_tree with string - two allocations.</figcaption>
+</figure>
+
+As you can see in Figure 1. - there's one allocation for the node within `radix_tree`
+and one for the string data itself. To overcome this, we proposed an `inline_string`,
+which keeps the actual data within the same allocation. It's a bit bigger, so it's
+optimized for writing, but it suffers when it comes to updating the string data.
+If the value is bigger than the existing string allocation - the whole node has to be
+re-allocated. In Figure 2. below, you can see that all the data is held together:
+metadata, key (right next to `inline_string` structure), and value.
+
+<figure class="image">
+  <img src="/assets/cpp_lessons_inline_string.png" alt="Radix tree with inline_string">
+  <figcaption>Figure 2. Radix_tree with inline_string - only one allocation.</figcaption>
+</figure>
+
+
+* **Using per-thread data**
+
+Very recently, we've described ([in the "Concurrency considerations" blog post][concurrency_blog])
+our approach to ensure data consistency. It's a very challenging task, especially if
+you consider concurrent modifications within a container full of records, with multiple
+threads accessing/modifying the same structure. Read the whole story for details,
+but long story short - per-thread data (implemented in libpmemobj-cpp as
+[enumerable_thread_specific][ptls_docs]) can save you from data consistency issues.
+
+
+* **Keeping locks on pmem is not a good idea**
+
+Since we keep all data for our application on persistent memory, one might think we should
+do that as well with locks. That's the wrong way. Each lock and unlock operation might end up
+with write to pmem. We measured the performance, and it turned out, this approach is slow
```

Comment:
I think it would be good to go into bit more detail here. The actual reason is that if a lock is accessed from different cores, the cache coherency mechanism might write the data to memory. (at least this is my understanding)